internal/states/encryption/encryption.go

```diff
@@ -0,0 +1,270 @@
+	//
+	// For indexed resources, please pass the index as follows:
+	//
+	//    encryptionconfig.Key("terraform_remote_state.foo[42]")
+	//    encryptionconfig.Key("terraform_remote_state.foo[test]")
+	//
+	// See encryptionconfig.Key for more details on the key format.
+	//
+	// When implementing this interface:
+	//
+	// - If the user provided no configuration, this function must return a flow that passes through the data
+	//   unmodified.
+	// - If the user only provided an environment configuration with the key encryptionconfig.KeyDefaultRemote, the
+	//   returned flow should use this configuration.
+	// - If the user provided a non-default HCL or environment configuration, these configurations should be merged
+	//   with the environment taking precedence. The default configuration should be ignored.
+	//
+	// Please note, the encryption and decryption fallback configuration may have separate configuration. This method
+	// should support this scenario to allow for encryption rollover.
+	//
+	// Tip: encryptionconfig.ConfigMap.Merge implements these precedence rules.
+	RemoteStateDatasource(configKey encryptionconfig.Key) (encryptionflow.StateFlow, error)
+
+	// StateFile returns an encryption flow suitable for encrypting the state file.
+	//
+	// When implementing this interface:
+	//
+	// - If the user provided no configuration, this function must return a flow that passes through the data
+	//   unmodified.
+	// - The default configuration is always ignored in this case because it is only the default for remote states.
+	// - If the user provided a non-default HCL or environment configuration, these configurations should be merged
+	//   with the environment taking precedence. The default configuration should be ignored.
+	//
+	// Please note, the encryption and decryption fallback configuration may have separate configuration. This method
+	// should support this scenario to allow for encryption rollover.
+	//
+	// Tip: encryptionconfig.ConfigMap.Merge implements these precedence rules.
+	StateFile() (encryptionflow.StateFlow, error)
+
+	// PlanFile returns an encryption flow suitable for encrypting the plan file.
+	//
+	// When implementing this interface:
+	//
+	// - If the user provided no configuration, this function must return a flow that passes through the data
+	//   unmodified.
+	// - The default configuration is always ignored in this case because it is only the default for remote states.
+	// - If the user provided a non-default HCL or environment configuration, these configurations should be merged
+	//   with the environment taking precedence. The default configuration should be ignored.
+	//
+	// Tip: encryptionconfig.ConfigMap.Merge implements these precedence rules.
+	PlanFile() (encryptionflow.PlanFlow, error)
+}
+
+type encryption struct {
+	encryptionConfigs         encryptionconfig.ConfigMap
+	decryptionFallbackConfigs encryptionconfig.ConfigMap
+	mutex                     *sync.Mutex
```

Comment:
Nit: as long as the encryption struct is passed by reference, there's no need to make this a pointer.